src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -2749,6 +2750,111 @@ std::string g_consoleFlagsOptionPrefix( "ai:console:" );
 const int g_logFlagsDefault = AI_LOG_ALL;
 const int g_consoleFlagsDefault = AI_LOG_WARNINGS | AI_LOG_ERRORS | AI_LOG_TIMESTAMP | AI_LOG_BACKTRACE | AI_LOG_MEMORY | AI_LOG_COLOR;
 
+void throwError( int errorCode )
+{
+	switch( errorCode )
+	{
+		case AI_ABORT :
+			throw IECore::Exception( "Render aborted" );
+		case AI_ERROR_NO_CAMERA :
+			throw IECore::Exception( "Camera not defined" );
+		case AI_ERROR_BAD_CAMERA :
+			throw IECore::Exception( "Bad camera" );
+		case AI_ERROR_VALIDATION :
+			throw IECore::Exception( "Usage not validated" );
+		case AI_ERROR_RENDER_REGION :
+			throw IECore::Exception( "Invalid render region" );
+		case AI_INTERRUPT :
+			throw IECore::Exception( "Render interrupted by user" );
+		case AI_ERROR_NO_OUTPUTS :
+			throw IECore::Exception( "No outputs" );
+		case AI_ERROR :
+			throw IECore::Exception( "Generic Arnold error" );
+	}
+}
+
+// Arnold's `AiRender()` function does exactly what you want for a batch render :
+// starts a render and returns when it is complete. But it is deprecated. Here we
+// jump through hoops to re-implement the behaviour using non-deprecated API.
+void renderAndWait()
+{
+
+	// Updated by `callback` to notify this thread when the render has
+	// completed.
+	struct Status {
+		std::mutex mutex;
+		std::condition_variable conditionVariable;
+		AtRenderStatus value = AI_RENDER_STATUS_NOT_STARTED;
+	} status;
+
+	// Called from one of the Arnold render threads to notify us of progress.
+	auto callback = []( void *voidStatus, AtRenderUpdateType updateType, const AtRenderUpdateInfo *updateInfo ) {
+
+		// We are required to return a new status for the render,
+		// following a table of values in `ai_render.h`.
+		AtRenderStatus newStatus = AI_RENDER_STATUS_FAILED;
+		switch( updateType )
+		{
+			case AI_RENDER_UPDATE_INTERRUPT :
+				newStatus = AI_RENDER_STATUS_PAUSED;
+				break;
+			case AI_RENDER_UPDATE_BEFORE_PASS :
+				newStatus = AI_RENDER_STATUS_RENDERING;
+				break;
+			case AI_RENDER_UPDATE_DURING_PASS :
+				newStatus = AI_RENDER_STATUS_RENDERING;
+				break;
+			case AI_RENDER_UPDATE_AFTER_PASS :
+				newStatus = AI_RENDER_STATUS_RENDERING;
+				break;
+			case AI_RENDER_UPDATE_IMAGERS :
+				// Documentation doesn't state the appropriate
+				// return value, so this is a guess.
+				newStatus = AI_RENDER_STATUS_RENDERING;
+				break;
+			case AI_RENDER_UPDATE_FINISHED :
+				newStatus = AI_RENDER_STATUS_FINISHED;
+				break;
+			case AI_RENDER_UPDATE_ERROR :
+				newStatus = AI_RENDER_STATUS_FAILED;
+				break;
+			// No `default` clause so that the compiler will warn us
+			// when new AtRenderUpdateType values are added.
+		}
```

Comment:
Wow, this is truly terrible.  But if it's what SolidAngle requires, then sure I guess.

Reply:
I should probably confess that there _is_ a simpler way :

```
AiRenderBegin( ... );
while( AiRenderGetStatus()` != AI_RENDER_STATUS_FINISHED && AiRenderGetStatus()` != AI_RENDER_STATUS_FAILED )
{
    std::this_thread::sleep_for( 100ms ); // What's the right duration to sleep for?
}
```

I was being a bit bloody minded in demonstrating just how many hoops you have to jump through to use the new API as intended. Solid Angle do seem to agree that they should have a single-line blocking batch render function though, so hopefully one will reappear in future. With that in mind, the other alternative is to use `pragma` to disable warnings for the old call.